### PR TITLE
Java API

### PR DESCRIPTION
**Pull request by hntd187**
After roughly 4000 years, this is finished...again.

All functionality should be there, same as Scala.

New Job Types are
- `JSparkJob`
- `JSqlJob`
- `JHiveJob`
- `JStreamingJob`

New Context Factories are
- `JavaSparkContextFactory`
- `JavaSqlContextFactory`
- `JavaHiveContextFactory`
- `JavaStreamingContextFactory`

An example Java Job would be written like this

```java
package spark.jobserver;

import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import org.apache.spark.api.java.JavaSparkContext;
import spark.jobserver.japi.JSparkJob;
import spark.jobserver.api.JobEnvironment;

public class JavaHelloWorldJob implements JSparkJob<String> {

    public String run(JavaSparkContext sc, JobEnvironment runtime, Config data) {
        return "Hi!";
    }

    public Config verify(JavaSparkContext sc, JobEnvironment runtime, Config config) {
        return ConfigFactory.empty();
    }
}
```

Everything is wrapped in `JavaJob` which is an interface to the SparkJobBase, but obviously making everything java friendly which produces the code above.

```scala
package spark.jobserver.japi

import com.typesafe.config.Config
import org.scalactic._
import spark.jobserver.api._

import scala.reflect.ClassTag

case class JavaValidationException(reason: String) extends Exception with ValidationProblem

case class JavaJob[R: ClassTag, CX: ClassTag](job: BaseJavaJob[R, CX]) extends SparkJobBase {

  import spark.jobserver.japi.{JavaValidationException => JVE}

  override type C = CX
  override type JobData = Config
  override type JobOutput = R

  def runJob(sc: CX, runtime: JobEnvironment, data: Config): R = {
    job.run(sc, runtime, data)
  }

  def validate(sc: CX, runtime: JobEnvironment, config: Config): Config Or Every[ValidationProblem] = {
    job.verify(sc, runtime, config) match {
      case j : JVE => Bad(One(j))
      case t : Throwable => Bad(One(JVE(t.getMessage)))
      case c : Config => Good(c)
    }
  }
}
```

This layer allows JavaJobs to basically run as if they are Scala jobs internally making life a lot easier. Now Java jobs just implement their given interface and then use the proper context factory to be provided the right context. 

There are probably a good bit of style or strangeness in here so feel free to rip apart my style.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -153,24 +154,24 @@ lazy val dockerSettings = Seq(
...
+      )
       volume("/database")
       entryPoint("app/server_start.sh")
     }
   },
   imageNames in docker := Seq(
     sbtdocker.ImageName(namespace = Some("velvia"),
-                        repository = "spark-jobserver",
-                        tag = Some(s"${version.value}.mesos-${mesosVersion.split('-')(0)}.spark-$sparkVersion.scala-${scalaBinaryVersion.value}"))
+      repository = "spark-jobserver",
```
Comment (velvia) [inline build.sbt]: I personally prefer the previous spacing, but no biggie

File: job-server-api/src/main/scala/spark/jobserver/JobCache.scala [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
+
+import scala.language.existentials
 
 trait BinaryJobInfo
 
 case class JobJarInfo(constructor: () => api.SparkJobBase,
                       className: String,
                       jarFilePath: String) extends BinaryJobInfo
 
-//For python jobs, there is no class loading or constructor required.
```
Comment (velvia) [inline job-server-api/src/main/scala/spark/jobserver/JobCache.scala]: Should we keep this comment?

File: job-server-extras/src/main/scala/spark/jobserver/context/JavaSqlContextFactory.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                         classPath: String,
+                         jobCache: JobCache): J Or LoadingError = {
+
+    val j = try {
+      jobCache.getJavaJob(appName, uploadTime, classPath)
+    } catch {
+      case _: ClassNotFoundException => return Bad(JobClassNotFound)
+      case err: Exception => return Bad(JobLoadError(err))
+    }
+    if (j.job.isInstanceOf[JStreamingJob[_]]) {
```
Comment (velvia) [inline job-server-extras/src/main/scala/spark/jobserver/context/JavaSqlContextFactory.scala]: It seems these ContextFactories are mostly the same except for this isInstanceOf....  maybe could be abstracted out.  OTOH we might get rid of these soon due to Spark 2.0 changes/upgrade.  Maybe just add a comment?

**Comment by velvia** [comment]
@hntd187 good job!   Looks great to me.   Do we need to update the README?   I'm also not sure why the code coverage numbers dropped.

I say let's merge it to get the Java API upgraded and then we can release 0.7.0.

File: job-server-api/src/main/scala/spark/jobserver/japi/JavaJob.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  override type JobOutput = R
+
+  def runJob(sc: CX, runtime: JobEnvironment, data: Config): R = {
+    job.run(sc, runtime, data)
+  }
+
+  def validate(sc: CX, runtime: JobEnvironment, config: Config): Config Or Every[ValidationProblem] = {
+    job.verify(sc, runtime, config) match {
+      case j : JVE => Bad(One(j))
+      case t : Throwable => Bad(One(JVE(t.getMessage)))
```
Comment (zeitos) [inline job-server-api/src/main/scala/spark/jobserver/japi/JavaJob.scala]: We will be loosing some important stack trace information here, consider adding t `JVE(t.getMessage, t)`

**Comment by velvia** [comment]
@hntd187 thanks, going to merge this.

**Comment by liweys** [comment]
Does it still work?

```
RESPONSE(code=500) {
  "status": "JOB LOADING FAILED",
  "result": {
    "message": "demo.Job4 cannot be cast to spark.jobserver.api.SparkJobBase",
    "errorClass": "java.lang.ClassCastException",
    "stack": ["spark.jobserver.context.ScalaContextFactory$class.loadAndValidateJob(SparkContextFactory.scala:87)", "spark.jobserver.context.DefaultSparkContextFactory.loadAndValidateJob(SparkContextFactory.scala:139)", "spark.jobserver.JobManagerActor.startJobInternal(JobManagerActor.scala:245)", "spark.jobserver.JobManagerActor$$anonfun$wrappedReceive$1.applyOrElse(JobManagerActor.scala:171)", "scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)", "spark.jobserver.common.akka.ActorStack$$anonfun$receive$1.applyOrElse(ActorStack.scala:33)", "scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)", "spark.jobserver.common.akka.Slf4jLogging$$anonfun$receive$1$$anonfun$applyOrElse$1.apply$mcV$sp(Slf4jLogging.scala:25)", "spark.jobserver.common.akka.Slf4jLogging$class.spark$jobserver$common$akka$Slf4jLogging$$withAkkaSourceLogging(Slf4jLogging.scala:34)", "spark.jobserver.common.akka.Slf4jLogging$$anonfun$receive$1.applyOrElse(Slf4jLogging.scala:24)", "scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)", "spark.jobserver.common.akka.ActorMetrics$$anonfun$receive$1.applyOrElse(ActorMetrics.scala:23)", "akka.actor.Actor$class.aroundReceive(Actor.scala:484)", "spark.jobserver.common.akka.InstrumentedActor.aroundReceive(InstrumentedActor.scala:8)", "akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)", "akka.actor.ActorCell.invoke(ActorCell.scala:495)", "akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)", "akka.dispatch.Mailbox.run(Mailbox.scala:224)", "akka.dispatch.Mailbox.exec(Mailbox.scala:234)", "scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)", "scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)", "scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)", "scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)"]
  }
}
```

**Comment by liweys** [comment]
Learnt from another thread it must be used together with JavaContextFactory,   
can this limitation be removed?

**Comment by hntd187** [comment]
It's not a limitation. The Java Context is separate from the Scala one, you shouldn't switch the usage of them. So unless Spark unifies and removes the Java context this probably won't go away. Also for these kinds of suggestions please make a new issue instead of digging up this old pull.

**Pull request merged**